hammers/scripts/gpu_lease_stacking.py

```diff
@@ -0,0 +1,145 @@
+# coding: utf-8
+'''
+.. code-block:: bash
+
+    gpu-lease-stacking {info, delete}
+
+Reclaims GPU nodes from leases that violate terms of use.
+
+* ``info`` to just display leases or actuall delete them with ``delete``
+'''
+import sys
+import argparse
+from pprint import pprint
+from datetime import datetime
+
+from hammers import MySqlArgs, osapi, query
+from hammers.slack import Slackbot
+from hammers.osrest.blazar import lease_delete
+
+
+def find_stacked_leases(leases):
+    """Return list of only the leases stacked on each other."""
+    stacked = []
+
+    for i in range(len(leases)):
+
+        start_date = leases[i][1]
+        end_date = leases[i][2]
+
+        if i > 0:
+            last_end_date = leases[i - 1][2]
+        else:
+            last_end_date = datetime.min
+
+        if i < len(leases) - 1:
+            next_start_date = leases[i + 1][1]
+        else:
+            next_start_date = datetime.max
+
+        stacked_previous = (start_date - last_end_date).days < 1
+        stacked_next = (next_start_date - end_date).days < 1
+
+        if stacked_previous or stacked_next:
+            stacked.append(leases[i])
+
+    return stacked
+
+
+def reaper(db, auth, describe=False, quiet=False):
+    """Delete stacked leases on gpu nodes."""
+    user_gpu_leases = {}
+
+    for row in query.gpu_leases(db):
+        user_id = row['user_id']
+        node_id = row['node_id']
+
+        if user_id not in user_gpu_leases.keys():
+            user_gpu_leases[user_id] = {}
+
+        if node_id not in user_gpu_leases[user_id].keys():
+            user_gpu_leases[user_id][node_id] = []
+
+        user_gpu_leases[user_id][node_id].append(
+            (row['lease_id'], row['start_date'], row['end_date']))
+
+    leases_to_delete = []
+    for user_id in user_gpu_leases.keys():
+        for node_id in user_gpu_leases[user_id].keys():
+            leases = list(set(user_gpu_leases[user_id][node_id]))
+            leases = list(sorted(leases, key=lambda x: x[1]))
+            stacked_leases = find_stacked_leases(leases)
+
+            leases_to_delete.extend(stacked_leases[1:])
```

Comment:
This `1` should probably be pulled out into a constant like `LEASES_ALLOWED` or something to make it more clear how it's working.